pjnath/src/pjnath/stun_session.c

```diff
@@ -507,6 +507,9 @@ static pj_status_t stun_tsx_on_send_msg(pj_stun_client_tsx *tsx,
     pj_status_t status;
 
     tdata = (pj_stun_tx_data*) pj_stun_client_tsx_get_data(tsx);
+    if (!tdata)
+        return PJ_EGONE;
+     
```

Comment:
Looks like a race condition happened. What if the context switching happens after the `tdata` validation (just added by this patch) instead of before?

Reply:
The method is called within a grup lock
https://github.com/pjsip/pjproject/blob/9a7d4d17202fb7821f3fbf1736c4f1468bdacb36/pjnath/src/pjnath/stun_transaction.c#L347-L354

and 

https://github.com/pjsip/pjproject/blob/9a7d4d17202fb7821f3fbf1736c4f1468bdacb36/pjnath/src/pjnath/stun_session.c#L145-L163

It shouldn't context switch within the group lock.

Reply:
Ah okay.

Just curious, it seems that theoritically they can be using different locks, one is tsx's and the other is sess's. So, what if the tdata query & check is moved down to inside the sess's lock, will it improve safety or even introduce some risk?

Reply:
We can't get the sess's lock without getting the tdata reference first. In this case we can't move the query/check